pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,164 @@
+type node struct {
+	name     string
+	tags     []string
+	children []*node
+}
+
+// String helps in pretty-printing a tree
+func (root *node) String() string {
+	tree := ""
+	tree += root.name
+
+	for _, n := range root.children {
+		tree += "(" + n.String() + ")"
+	}
+
+	return tree
+}
+
+const longDescription = `Experimental command
+
+Output build dependencies of a specific image repository.
+
+Examples:
+
+    $ openshift ex build-chain [image-repository]
+`
+
+func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s [image-repository]", name),
+		Short: "Output build dependencies of a specific image repository",
+		Long:  longDescription,
+		Run: func(cmd *cobra.Command, args []string) {
+			imageRepoName := cmdutil.GetFlagString(cmd, "imagerepo")
+			if imageRepoName == "" {
+				usageError(cmd, "Must pass an image repository name with the --imagerepo (-i) flag")
+			}
+
+			c, _, err := f.Clients(cmd)
+			checkErr(err)
+
+			namespace, err := f.DefaultNamespace(cmd)
+			checkErr(err)
+
+			// Validate the specified image repository
+			if !isValid(imageRepoName, c, namespace) {
+				checkErr(fmt.Errorf("invalid image repository: %s", imageRepoName))
+			}
+
+			// Get all build configurations and look for those
+			// that reference the specified image repository
+			buildConfigList, err := c.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+			checkErr(err)
+
+			root := findRepoDeps(imageRepoName, buildConfigList.Items)
+			fmt.Println(root)
+			// imgRoot := repoToImage(c, namespace, imageRepoName, root)
+		},
+	}
+
+	cmd.Flags().StringP("imagerepo", "i", "", "Image repository for which to output build dependencies")
+	return cmd
+}
+
+// isValid checks if the given image repository is valid
+func isValid(name string, c *client.Client, namespace string) bool {
+	// TODO: Validate image repository via buildConfigs?
+	repoList, err := c.ImageRepositories(namespace).List(labels.Everything(), labels.Everything())
+	checkErr(err)
+
+	for _, repo := range repoList.Items {
+		if name == repo.Name {
+			if errs := val.ValidateImageRepository(&repo); len(errs) == 0 {
+				return true
+			}
+			break
+		}
+	}
+
+	return false
+}
+
+// findRepoDeps accepts an image repository name and a list of build
+// configurations and returns a dependency tree of the specified
+// image repository along with its tags
+func findRepoDeps(from string, buildConfigList []buildapi.BuildConfig) *node {
+	root := &node{
+		name: from,
+	}
+
+	// Search all build configurations in order to find the image
+	// repositories depending on the specified image repository
+	for _, cfg := range buildConfigList {
+		for _, trigger := range cfg.Triggers {
+			if from == trigger.ImageChange.From.Name {
+				// Search recursively all children for their dependency trees
+				// TODO: Check for both outputs ? (To & Tag or DockerImageReference)
```

Comment:
We're interested in both, but they mean different things.

----- Original Message -----
> > +
> > +// findRepoDeps accepts an image repository name and a list of build
> > +// configurations and returns a dependency tree of the specified
> > +// image repository along with its tags
> > +func findRepoDeps(from string, buildConfigList []buildapi.BuildConfig)
> > *node {
> > +	root := &node{
> > +		name: from,
> > +	}
> > +
> > +	// Search all build configurations in order to find the image
> > +	// repositories depending on the specified image repository
> > +	for _, cfg := range buildConfigList {
> > +		for _, trigger := range cfg.Triggers {
> > +			if from == trigger.ImageChange.From.Name {
> > +				// Search recursively all children for their dependency trees
> > +				// TODO: Check for both outputs ? (To & Tag or DockerImageReference)
> 
> I would think we're only interested in output.to, @bparees ?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26134297